design/how-does-data-distribution-work.md

```diff
@@ -0,0 +1,83 @@
+
+### How to move keys?
+
+A key range is a shard. A shard is the minimum unit of moving data. The storage server’s ownership of a shard -- which SS owns which shard -- is stored in system keyspace *serverKeys *(\xff/serverKeys/) and *keyServers* (\xff/keyServers/). To simplify the explanation, we refer to the storage server’s ownership of a shard as a shard’s ownership. 
+
+Shard’s ownership is used in transaction systems (proxy and tLogs) to route mutations to tLogs and storage servers. When proxy receives a mutation, it uses the shard’s ownership to decide which *k* tLogs receive the mutation, assuming *k* is the replias factor. When a storage server pulls mutations from tLogs, it uses the shard’s ownership to decide which shards the SS is responsible for and which tLog the SS should pull the data.
+
+Shard’s ownership must be consistent across transaction systems and SSes, so that mutations can be correctly routed to SSes. Moving keys from a SS to another requires changing the shard’s ownership under ACID property. The ACID property is achieved by using FDB transactions to change the *serverKeys *(\xff/serverKeys/) and *keyServers* (\xff/keyServers/). The mutation on the *serverKeys *and* keyServers *will be categorized as private mutations in transaction system. Compared to normal mutation, the private mutations will change the transaction state store (txnStateStore) that maintains the *serverKeys *and* keyServers *for transaction systems (proxy and tLog) when it arrives on each transaction component (e.g., tLog). Because mutations are processed in total order with the ACID guarantees, the change to the txnStateStore will be executed in total order on each node and the change on the shard’s ownership will also be consistent.
+
+The data movement from one server (called source server) to another (called destination server) has four steps: (1) DD adds the destination server as the shard’s new owner; (2) the destination server will issue transactions to read the shard range and write the key-value pairs back. The key-value will be routed to the destination server and saved in the server’s storage engine; (3) DD removes the source server from the shard’s ownership; (4) DD removes the shard’s information owned by the source server from the server’s team information (i.e., *shardsAffectedByTeamFailure*). 
```

Comment:
```suggestion
The data movement from one server (called source server) to another (called destination server) has four steps: (1) DD adds the destination server as the shard’s new owner; (2) the destination server will issue transactions to read the shard range and write the key-value pairs back. The key-value pairs will be routed to the destination server and saved in the server’s storage engine; (3) DD removes the source server from the shard’s ownership; (4) DD removes the shard’s information owned by the source server from the server’s team information (i.e., *shardsAffectedByTeamFailure*). 
```

Reply:
Consider using a new line for each bullet. Also for (3), we should specify this is done by modifying the system keyspace